fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+        // /// different threads.
+        // fn post(&self);
+        // /// Returns the fd to read from for notification.
+        // fn read_fd(&self) -> i32;
+        // /// If an event is signalled, consume it; otherwise return.
+        // /// This does not block. This retries on `EINTR`.
+        // fn try_consume(&self) -> bool;
+        #[rust_name = "new_fd_event_signaller"]
+        fn ffi_new_fd_event_signaller_t() -> UniquePtr<FdEventSignaller>;
+    }
+    // unsafe extern "C++" {
+    //     include!("fds.h");
+    //
+    //     /// A helper class for managing and automatically closing a file descriptor.
+    //     #[rust_name = "AutoCloseFd"]
+    //     type autoclose_fd_t; // = crate::ffi::autoclose_fd_t;
+    //     /// Closes the fd if not already closed
+    //     fn close(self: Pin<&mut AutoCloseFd>);
+    //     /// Returns the internal/raw fd.
+    //     fn fd(&self) -> i32;
+    //     /// Returns the fd, transferring ownership to the caller
+    //     fn acquire(self: Pin<&mut AutoCloseFd>) -> i32;
+    //     /// Resets to a new fd, taking ownership.
+    //     fn reset(self: Pin<&mut AutoCloseFd>, fd: i32);
+    //     /// Returns `true` if this `AutoCloseFd` instance has a valid fd (`>= 0`).
+    //     #[rust_name = "is_valid"]
+    //     fn valid(&self) -> bool;
+    // }
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_id_t"]
+        type FdMonitorItemId;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_t"]
+        type FdMonitorItem;
+
+        #[cxx_name = "make_fd_monitor_item_t"]
+        fn new_fd_monitor_item_ffi(fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> Box<FdMonitorItem>;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_t"]
+        type FdMonitor;
+
+        #[cxx_name = "make_fd_monitor_t"]
+        fn new_fd_monitor_ffi() -> Box<FdMonitor>;
+
+        #[cxx_name = "add_item"]
+        fn add_item_ffi(&mut self, fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> u64;
+
+        #[cxx_name = "poke_item"]
+        fn poke_item_ffi(&self, item_id: u64);
+
+        #[cxx_name = "add"]
+        pub fn add_ffi(&mut self, item: Box<FdMonitorItem>) -> u64;
+    }
+}
+
+// TODO: Remove once we're no longer using the FFI variants of FdEventSignaller and AutoCloseFd.
+unsafe impl Sync for FdEventSignaller {}
+unsafe impl Send for FdEventSignaller {}
+// unsafe impl Send for AutoCloseFd {}
+// unsafe impl Sync for AutoCloseFd {}
+
+// TODO: Remove once we're no longer using the FFI variant of AutoCloseFd.
+impl Unpin for FdMonitorItem {}
+
+/// Each item added to fd_monitor_t is assigned a unique ID, which is not recycled. Items may have
+/// their callback triggered immediately by passing the ID. Zero is a sentinel.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct FdMonitorItemId(u64);
```

Comment:
probably no need for `repr(transparent)`. C++ won't write this directly, right?

Reply:
I was going to transmute or pointer cast originally.